pkg/proxy/backend/authenticator.go

```diff
@@ -24,74 +24,116 @@ import (
+	pnet.ClientMultiResults | pnet.ClientPluginAuth | pnet.ClientConnectAttrs | pnet.ClientDeprecateEOF
 
 // Authenticator handshakes with the client and the backend.
 type Authenticator struct {
-	backendTLSConfig *tls.Config
-	dbname           string // default database name
-	serverAddr       string
-	user             string
-	attrs            []byte // no need to parse
-	capability       uint32 // client capability
-	collation        uint8
+	backendTLSConfig            *tls.Config
+	supportedServerCapabilities pnet.Capability
+	dbname                      string // default database name
+	serverAddr                  string
+	user                        string
+	attrs                       []byte // no need to parse
+	capability                  uint32 // client capability
+	collation                   uint8
 }
 
 func (auth *Authenticator) String() string {
 	return fmt.Sprintf("user:%s, dbname:%s, capability:%d, collation:%d",
 		auth.user, auth.dbname, auth.capability, auth.collation)
 }
 
-func (auth *Authenticator) handshakeFirstTime(clientIO, backendIO *pnet.PacketIO, frontendTLSConfig, backendTLSConfig *tls.Config) error {
+func (auth *Authenticator) handshakeFirstTime(logger *zap.Logger, clientIO, backendIO *pnet.PacketIO, salt []byte, frontendTLSConfig, backendTLSConfig *tls.Config, proxyProtocol bool) error {
+	clientIO.ResetSequence()
 	backendIO.ResetSequence()
-	// Read initial handshake packet from the backend.
-	serverPkt, serverCapability, err := auth.readInitialHandshake(backendIO)
-	if serverPkt != nil {
-		if writeErr := clientIO.WritePacket(serverPkt, true); writeErr != nil {
-			return writeErr
+
+	serverCapability := auth.supportedServerCapabilities
+	if frontendTLSConfig == nil {
+		serverCapability ^= pnet.ClientSSL
+	}
+	if err := clientIO.WriteInitialHandshake(serverCapability.Uint32(), salt, mysql.AuthCachingSha2Password); err != nil {
+		return errors.WithStack(err)
+	}
+	pkt, isSSL, err := clientIO.ReadSSLRequestOrHandshakeResp()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	clientCapability := pnet.Capability(binary.LittleEndian.Uint32(pkt))
+	if isSSL {
+		if serverCapability&pnet.ClientSSL == 0 {
+			return clientIO.WriteErrPacket(mysql.NewErr(mysql.ErrAbortingConnection, "should not go here", nil))
 		}
+		if _, err = clientIO.UpgradeToServerTLS(frontendTLSConfig); err != nil {
+			return errors.WithStack(err)
+		}
+		pkt, isSSL, err = clientIO.ReadSSLRequestOrHandshakeResp()
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		if isSSL {
+			return errors.WithStack(errors.New("expect handshake resp"))
+		}
+	} else {
+		binary.LittleEndian.PutUint32(pkt, (clientCapability | pnet.ClientSSL).Uint32())
+	}
+	if cc := clientCapability & serverCapability; clientCapability^cc != 0 {
+		logger.Info("unsupported capability from the frontend", zap.Stringer("common", cc), zap.Stringer("frontend", clientCapability^cc), zap.Stringer("proxy", serverCapability^cc))
```

Comment:
This is not right. Some capabilities are not really capabilities, they are just client options.
E.g. the default server capability doesn't contain `ClientNoSchema` but the client may send it to specify that the handshake response doesn't contain a schema. The server still supports it.
That's why I added a `supportedServerCapabilities` to specify what the proxy supports.

Reply:
Then it is conflicted with the official mysql document:

> The client should only announce the capabilities in the [Protocol::HandshakeResponse](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_packets_protocol_handshake_response.html) that it has in common with the server.

Server should at least warn about uncommon flags. Capability negotiation works by assuming common capabilities.

TBH, I don't think it is good idea to be slient about those unsupported capabilities. You can see that TiDB does not support `ClientNoSchema`, too. But TiDB will be slient about it, which means it only works if clients respect the mysql protocol strictly.

```
[xhe@M1PRO a]$ rg ClientNoSchema ../pingcap/tidb
../pingcap/tidb/parser/mysql/const.go
140:	ClientNoSchema

No other occurrences
```

Reply:
The MySQL document is not always right. See  https://github.com/pingcap/tidb/blob/master/server/conn.go#L500

Reply:
Besides, the proxy can support `CLIENT_ODBC` because it has no special behaviors.
 https://dev.mysql.com/doc/internals/en/capability-flags.html#flag-CLIENT_ODBC

Reply:
> Besides, the proxy can support CLIENT_ODBC because it has no special behaviors.

Then we should add it to the supported set.

> The MySQL document is not always right. 

It is that mysql-client is wrong. We should respect the spec as much as possible.

>  You cannot test all clients.

If you feel it is too strict to forbid all cases. We can just report the difference without really disconnect it.